zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs

```diff
@@ -188,23 +255,90 @@ fn checked_add_transaction_weighted_random(
     //
     // Unpaid actions are always zero for transactions that pay the conventional fee,
     // so the unpaid action check always passes for those transactions.
-    if candidate_tx.transaction.size <= *remaining_block_bytes
-        && candidate_tx.legacy_sigop_count <= *remaining_block_sigops
-        && candidate_tx.unpaid_actions <= *remaining_block_unpaid_actions
-    {
-        selected_txs.push(candidate_tx.clone());
-
-        *remaining_block_bytes -= candidate_tx.transaction.size;
-        *remaining_block_sigops -= candidate_tx.legacy_sigop_count;
-
-        // Unpaid actions are always zero for transactions that pay the conventional fee,
-        // so this limit always remains the same after they are added.
-        *remaining_block_unpaid_actions -= candidate_tx.unpaid_actions;
+    if candidate_tx.try_update_block_limits(
+        remaining_block_bytes,
+        remaining_block_sigops,
+        remaining_block_unpaid_actions,
+    ) {
+        let selected_tx_id = candidate_tx.transaction.id.mined_id();
+        selected_txs.push((0, candidate_tx));
+
+        for dependent_candidate_tx_ids_by_level in
+            mempool_tx_deps.all_dependents_leveled(&selected_tx_id)
+        {
+            for dependent_candidate_tx_id in &dependent_candidate_tx_ids_by_level {
+                let mempool_tx_deps = mempool_tx_deps.dependencies();
+                let candidate_tx_deps = mempool_tx_deps.get(dependent_candidate_tx_id);
+
+                if has_direct_dependencies(candidate_tx_deps, selected_txs) {
```

Comment:
If anyone has any ideas on how to address this, there's an issue here that I'm not sure how to fix this yet where tx3 won't be selected if tx1 is independent, and tx2 depends on tx1, but tx3 depends on both tx1 and tx2, because it's checked before tx2 is selected. 

I thought about storing those in another collection to try selecting them after the rest of the dependent txs, but that still won't work if there's a tx4 that depends on tx1 and tx3.